packages/oui-dual-list/src/dual-list.controller.js

```diff
@@ -0,0 +1,247 @@
+        this.sourceOpenIconStyle = this.displayBlockStyle;
+        this.sourceContentStyle = this.displayFlexStyle;
+    }
+
+    onSourceContentOpen () {
+        this.sourceCloseIconStyle = this.displayBlockStyle;
+        this.sourceOpenIconStyle = this.displayNoneStyle;
+        this.sourceContentStyle = this.displayNoneStyle;
+    }
+
+    getProperty (item) {
+        if (!this.property) {
+            return item;
+        }
+        return this.property.split(".").reduce((prev, curr) => prev ? prev[curr] : undefined, item);
+    }
+
+    isLoading (item) {
+        const uniqueName = this.getProperty(item);
+        if (this.loadingMap[uniqueName]) {
+            return true;
+        }
+        return false;
+    }
+
+    setLoading (item, state) {
+        const uniqueName = this.getProperty(item);
+        this.loadingMap[uniqueName] = state;
+    }
+
+    loadSourceList () {
+        if (this.sourceListLoading) {
+            return this.$q.reject(false);
+        }
+        this.sourceListLoading = true;
+        return this.$q.when(this.sourceList)
+            .then(items => {
+                this.sourceList = items.data ? items.data : items;
+            })
+            .finally(() => {
+                this.sourceListLoading = false;
+            });
+    }
+
+    loadtargetList () {
+        if (this.targetListLoading) {
+            return this.$q.reject(false);
+        }
+        this.targetListLoading = true;
+        return this.$q.when(this.targetList)
+            .then(items => {
+                this.targetList = items.data ? items.data : items;
+            })
+            .finally(() => {
+                this.targetListLoading = false;
+            });
+    }
+
+    add (index, item) {
+        if (this.isLoading(item)) {
+            return;
+        }
+        this.sourceList.splice(index, 1);
+        this.targetList.push(item);
+        if (this.onAdd) {
+            this.setLoading(item, true);
+            this.onAdd({ items: [item] })
+                .catch(() => {
+                    const newIndex = this.targetList.indexOf(item);
+                    this.targetList.splice(newIndex, 1);
+                    this.sourceList.push(item);
+                })
+                .finally(() => this.setLoading(item, false));
+        }
+    }
+
+    remove (index, item) {
+        if (this.isLoading(item)) {
+            return;
+        }
+        this.targetList.splice(index, 1);
+        this.sourceList.push(item);
+        if (this.onRemove) {
+            this.setLoading(item, true);
+            this.onRemove({ items: [item] })
+                .catch(() => {
+                    const newIndex = this.sourceList.indexOf(item);
+                    this.sourceList.splice(newIndex, 1);
+                    this.targetList.push(item);
+                })
+                .finally(() => this.setLoading(item, false));
+        }
+    }
+
+    addAll () {
```

Comment:
I really think this method (and `removeAll`) can be simplified by merging the source array (filtered or not) to the other, instead of pushing each item at a time in the other. Maybe, is there a reason to do it this way ?

Reply:
Yes, there is a reason. All items can not be moved blindly when the user clicks add all link. Items which are loading should not be added. That's why I loop through all items and pick items which are not in the loading state.